Tests/UnitTests/Networking/HTTPClientTests.swift

```diff
@@ -1664,6 +1666,210 @@ final class HTTPClientTests: BaseHTTPClientTests<MockETagManager> {
+        let retryAfterMilliseconds = 100
+
+        let httpURLResponse = HTTPURLResponse(
+            url: URL(string: "api.revenuecat.com")!,
+            statusCode: HTTPStatusCode.tooManyRequests.rawValue,
+            httpVersion: nil,
+            headerFields: [
+                HTTPClient.ResponseHeader.retryAfter.rawValue: "\(retryAfterMilliseconds)"
+            ]
+        )!
+
+        let backoffPeriod = self.client.calculateRetryBackoffTime(forResponse: httpURLResponse, retryCount: 2)
+        expect(backoffPeriod).to(equal(TimeInterval(0.1)))  // 0.1s == 100ms
+    }
+
+    func testUsesDefaultBackoffIfServerProvidedBackoffMissing() {
+        let httpURLResponse = HTTPURLResponse(
+            url: URL(string: "api.revenuecat.com")!,
+            statusCode: HTTPStatusCode.tooManyRequests.rawValue,
+            httpVersion: nil,
+            headerFields: [:]
+        )!
+
+        let backoffPeriod = self.client.calculateRetryBackoffTime(forResponse: httpURLResponse, retryCount: 2)
+        expect(backoffPeriod).to(equal(TimeInterval(0.25)))
+    }
+
+    func testDefaultRetryBackoffPeriods() {
+        expect(
+            self.client.defaultExponentialBackoffTimeInterval(withRetryCount: 1)
+        ).to(equal(0))
+
+        expect(
+            self.client.defaultExponentialBackoffTimeInterval(withRetryCount: 2)
+        ).to(equal(0.25))
+
+        expect(
+            self.client.defaultExponentialBackoffTimeInterval(withRetryCount: 3)
+        ).to(equal(0.5))
+    }
+
+    func testPerformsAllRetriesIfAlwaysGetsRetryableStatusCode() throws {
+        var requestCount = 0
+
+        let host = try XCTUnwrap(HTTPRequest.Path.serverHostURL.host)
+        stub(condition: isHost(host)) { _ in
+            requestCount += 1
+            return .emptyTooManyRequestsResponse()
+        }
+
+        let request = HTTPRequest(method: .get, path: .mockPath)
+        let result = waitUntilValue { completion in
+            self.client.perform(request) { (response: EmptyResponse) in
+                completion(response)
+            }
+        }
+        expect(requestCount).to(equal(4)) // 1 original request + 3 retries
+
+        expect(result).toNot(beNil())
+        expect(result).to(beFailure())
+
+        let error = try XCTUnwrap(result?.error)
+        expect(error) == .errorResponse(
+            .init(code: .unknownError,
+                  originalCode: 0,
+                  message: nil),
+            .tooManyRequests
+        )
+        expect(error.isServerDown) == false
+        self.logger.verifyMessageWasLogged("Queued request GET /v1/subscribers/identify for retry in 0.0 seconds.")
+        self.logger.verifyMessageWasLogged("Queued request GET /v1/subscribers/identify for retry in 0.25 seconds.")
+        self.logger.verifyMessageWasLogged("Queued request GET /v1/subscribers/identify for retry in 0.5 seconds.")
+        self.logger.verifyMessageWasLogged("Request GET /v1/subscribers/identify failed all 3 retries.")
```

Comment:
these are quite dangerous in that the minute we change the logs we mess up tests. That'd be fine if we were trying to specifically test that we're logging, but here we're testing "Performs all retries if always gets retryable status code", so we should instead make sure to test for that. 

We test for request count to hit 4, maybe we could also mock the operation dispatcher as a separate test and ensure that it's getting the right delays in?

Reply:
Love this! I've moved the log assertions to their own tests and created separate test to verify that the correct delay times are being sent to the OperationDispatcher.